lib/Frontend/CachedDiagnostics.cpp

```diff
@@ -111,31 +115,36 @@ struct DiagnosticSerializer {
   llvm::Error serializeEmittedDiagnostics(llvm::raw_ostream &os);
 
   static llvm::Error
-  emitDiagnosticsFromCached(llvm::StringRef Buffer, SourceManager &SrcMgr,
-                            DiagnosticEngine &Diags,
-                            llvm::PrefixMapper &Mapper,
+  emitDiagnosticsFromCached(llvm::StringRef Buffer,
+                            DiagnosticEngine &Diags, llvm::PrefixMapper &Mapper,
+                            llvm::cas::ObjectStore &CAS,
                             const FrontendInputsAndOutputs &InAndOut) {
     // Create a new DiagnosticSerializer since this cannot be shared with a
-    // serialization instance.
-    DiagnosticSerializer DS(SrcMgr.getFileSystem(), Mapper);
-    DS.addInputsToSourceMgr(InAndOut);
+    // serialization instance. Using an empty in-memory file system as
+    // underlying file system because the replay logic should not touch file
+    // system.
+    auto FS = llvm::makeIntrusiveRefCnt<llvm::vfs::InMemoryFileSystem>();
+    DiagnosticSerializer DS(std::move(FS), Mapper, CAS);
     return DS.doEmitFromCached(Buffer, Diags);
   }
 
   SourceManager &getSourceMgr() { return SrcMgr; }
 
-  void addInputsToSourceMgr(const FrontendInputsAndOutputs &InAndOut) {
+  void addInputsToSourceMgr(SourceManager &SM,
+                            const FrontendInputsAndOutputs &InAndOut) {
     // Extract all the input file names so they can be added to the source
     // manager when replaying the diagnostics. All input files are needed even
     // they don't contain diagnostics because FileSpecificDiagConsumer need
     // has references to input files to find subconsumer.
     auto addInputToSourceMgr = [&](const InputFile &Input) {
-      if (Input.getFileName() != "-")
-        SrcMgr.getExternalSourceBufferID(remapFilePath(Input.getFileName()));
+      auto Path = remapFilePath(Input.getFileName());
+      SrcMgr.getExternalSourceBufferID(Path);
```

Comment:
Duplicated with next line

Reply:
Those are two different SourceMgr. The first one is the one used to replay, and the second one is the actual SourceMgr from the instance.